include/fiction/algorithms/simulation/sidb/quickexact.hpp

```diff
@@ -0,0 +1,506 @@
+                        dependent_cell_mode::VARIABLE);  // "false" allows that the charge state of the dependent cell
+                                                         // is automatically changed based on the new charge
+                                                         // distribution.
+                }
+
+                if (charge_lyt.is_physically_valid())
+                {
+                    charge_distribution_surface<Lyt> charge_lyt_copy{charge_lyt};
+                    result.charge_distributions.push_back(charge_lyt_copy);
+                }
+            }
+            // If the layout consists of only detected negatively-charged SiDBs.
+            else if (all_sidbs_in_lyt_without_negative_detected_ones.empty() && number_of_sidbs > 1)
+            {
+                charge_distribution_surface<Lyt> charge_lyt_copy{charge_lyt};
+                result.charge_distributions.push_back(charge_lyt_copy);
+            }
+
+            for (const auto& cell : detected_negative_sidbs)
+            {
+                layout.assign_cell_type(cell, Lyt::cell_type::NORMAL);
+            }
+        }
+
+        result.simulation_runtime = time_counter;
+
+        return result;
+    }
+
+  private:
+    /**
+     * Layout to simulate.
+     */
+    Lyt layout;
+    /**
+     * Charge distribution surface.
+     */
+    charge_distribution_surface<Lyt> charge_lyt{};
+    /**
+     * Parameters used for the simulation.
+     */
+    quickexact_params<Lyt> params{};
+    /**
+     * Indices of all SiDBs that are detected to be negatively charged in a physically valid layout.
+     */
+    std::vector<int64_t> detected_negative_sidb_indices{};
+    /**
+     * All SiDBs that are detected to be negatively charged in a physically valid layout.
+     */
+    std::vector<typename Lyt::cell> detected_negative_sidbs{};
+    /**
+     * All SiDBs of the layout but without the negatively-charged SiDBs.
+     */
+    std::vector<typename Lyt::cell> all_sidbs_in_lyt_without_negative_detected_ones{};
+    /**
+     * Collection of defects that are placed in addition to the SiDBs.
+     */
+    std::unordered_map<typename Lyt::cell, const sidb_defect> real_placed_defects{};
+    /**
+     * Number of SiDBs of the input layout.
+     */
+    uint64_t number_of_sidbs{};
+    /**
+     * Simulation results.
+     */
+    sidb_simulation_result<Lyt> result{};
+
+    /**
+     * This function conducts 2-state physical simulation (negative, neutral).
+     *
+     * @param charge_layout Initialized charge layout.
+     */
+    void two_state_simulation(charge_distribution_surface<Lyt>& charge_layout)
+    {
+        charge_layout.assign_base_number(2);
+        uint64_t current_charge_index  = 0;
+        uint64_t previous_charge_index = 0;
+
+        for (uint64_t i = 0; i <= charge_layout.get_max_charge_index(); i++)
+        {
+            current_charge_index = (i ^ (i >> 1u));  // gray code is used for the charge index.
```

Comment:
Unclear why this is Gray code. I would implement a `gray_code_iterator` that iterates over a range of integers in Gray code order.

Also, it is not documented here, why you use Gray code.